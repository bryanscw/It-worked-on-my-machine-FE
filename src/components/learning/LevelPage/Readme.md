### Example:

```js

import 'bootstrap/dist/css/bootstrap.css';
import 'bootstrap/dist/js/bootstrap.js';

import { METHODS } from '../../../redux/ducks/apiHelper.js';
import { withReduxRouter } from '../../../utils/mock.js';

const matchObject = {
    params: {
        topicId: 1,
        levelId: 1,
    }
}
const initialState = {
    authReducer: {
        user: {
            "name": "Bob"
        }
    },
    levelsReducer: {
        item: 
                    {
                        "id": 1,
                        "title": "Adding 1 to 100",
                        "description": "Small numbers.",
                        "topic":1,
                        "playable":true

                        },
        isLoading: {
            [METHODS.LIST]: false
        },
        hasLoaded: {
            [METHODS.LIST]: false
        }
    },
    questionsReducer: {
        items: [
                    {
                        "id": 1,
                        "questionText": "1 + 1",
                        "options": {
                            "0": 1,
                            "1": 2,
                        },
                        "answer": 1,
                    },
                    {
                        "id": 2,
                        "questionText": "1 + 2",
                        "options": {
                            "0": 1,
                            "1": 2,
                            "2": 3,
                        },
                        "answer": 2,
                    }
                ],
        isLoading: {
            [METHODS.LIST]: false
        },
        hasFailed: {
            [METHODS.LIST]: false
        }
    },
    learningMaterialsReducer: {
        items: [
                    {
                            "id": 1,
                            "title": "Basic addition",
                            "description": "Interesting video",
                            "link": "https://www.youtube.com/embed/Fe8u2I3vmHU",
                        }
                ],
        isLoading: {
            [METHODS.LIST]: false
        },
        hasFailed: {
            [METHODS.LIST]: false
        }
    },
    progressReducer: {
        crudlReducer: {
            item: {
                "complete":true
            },
            isLoading: {
            [METHODS.RETRIEVE]: false
        },
        hasFailed: {
            [METHODS.RETRIEVE]: false
        }

        }
    }
}

withReduxRouter(<LevelPage match={matchObject}/>, initialState)
```