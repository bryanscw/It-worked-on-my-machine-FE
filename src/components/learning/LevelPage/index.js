import React, { Component } from 'react';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons';

import Loader from '../../common/Loader';
import LearningMaterial from '../LearningMaterial';
import { retrieveLevel, selectPlayableLevel, selectLevelLoading, selectLevelFailed } from '../../../redux/ducks/levels';

export class LevelPage extends Component {
    constructor(props) {
        super(props);

        const topicId = parseInt(props.match.params.topicId);
        const levelId = parseInt(props.match.params.levelId);
        props.retrieveLevel(topicId, levelId);
    }
    
    render() {
        const {
            levelLoading,
            levelFailed,
            level,
        } = this.props;

        // TODO: Fetch this from API
        const viewedBefore = true;

        if (levelLoading)
            return <Loader />;

        if (levelFailed || !level)
            return <Redirect to="/not-found" />;

        return (
            <div className="container">
                <Link className="btn btn-light mb-2" to={`/topics/${level.topic.id}/`}>
                    <FontAwesomeIcon icon={faChevronLeft}/> Back to Topic Page
                </Link>
                <h1>{level.title}</h1>
                <div className="mb-4">
                    <Link className="btn btn-primary" to={`/topics/${level.topic.id}/levels/${level.id}/leaderboard`}>
                        View Leaderboard
                    </Link>
                </div>
                {
                    viewedBefore
                        ? <div>
                            <button className="btn btn-primary mb-4" type="button" data-toggle="collapse" data-target="#learningMaterialCollapse">
                                View Learning Material
                            </button>
                            <div className="collapse mb-4" id="learningMaterialCollapse">
                                <LearningMaterial levelId={level.id} />
                            </div>
                        </div>
                        : <div className="mb-4">
                            <h2>Learning Material</h2>
                            <LearningMaterial levelId={level.id} />
                        </div>
                }
                <Link className="btn btn-primary mb-4" to={`/topics/${level.topic.id}/levels/${level.id}/game`}>
                    Play Game
                </Link>
            </div>
        );
    }
}

const mapStateToProps = state => ({
    levelLoading: selectLevelLoading(state),
    levelFailed: selectLevelFailed(state),
    level: selectPlayableLevel(state),
});

const dispatchers = {
    retrieveLevel,
};

export default connect(mapStateToProps, dispatchers)(LevelPage);