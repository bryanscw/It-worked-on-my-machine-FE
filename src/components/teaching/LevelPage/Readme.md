### Example:

```js
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { MemoryRouter as Router } from 'react-router-dom';
import 'bootstrap/dist/css/bootstrap.css';
import 'bootstrap/dist/js/bootstrap.js';

import { METHODS } from '../../../redux/ducks/apiHelper.js';
import { withReduxRouter } from '../../../utils/mock.js';

const matchObject = {
    params: {
        topicId: 1,
        levelId: 1,
    }
}
const initialState = {
    authReducer: {
        user: {
            "name": "Bob"
        }
    },
    topicsReducer: {
        items: [
                    {
                        "id": 1,
                        "questionText": "1 + 1",
                        "options": {
                            "0": 1,
                            "1": 2,
                        },
                        "answer": 1,
                    }
                    ],
        isLoading: {
            [METHODS.LIST]: false
        },
        hasLoaded: {
            [METHODS.LIST]: false
        }
    }
}

withReduxRouter(<LevelPage match={matchObject}/>, initialState).renderedComponent
```