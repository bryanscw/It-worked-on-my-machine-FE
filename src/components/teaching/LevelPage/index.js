import React, { Component } from 'react';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons';
import PropTypes from 'prop-types'

import Loader from '../../common/Loader';
import LearningMaterialList from '../LearningMaterialList';
import QuestionsList from '../QuestionsList';
import { retrieveLevel, publishLevel, selectLevel, selectLevelLoading, selectLevelFailed } from '../../../redux/ducks/levels';

/**
 * This component displays the level page for a teacher. It contains the QuestionsList and LearningMaterialList components.
 */
export class LevelPage extends Component {
    componentDidMount() {
        const topicId = parseInt(this.props.match.params.topicId);
        const levelId = parseInt(this.props.match.params.levelId);
        this.props.retrieveLevel(topicId, levelId);
    }

    onPublish = () => {
        const topicId = parseInt(this.props.match.params.topicId);
        const levelId = parseInt(this.props.match.params.levelId);
        this.props.publishLevel(topicId, levelId);
    }
    
    render() {
        const {
            levelLoading,
            levelFailed,
            level,
        } = this.props;

        if (levelLoading)
            return <Loader />;

        if (levelFailed || !level)
            return <Redirect to="/not-found" />;
        
        return (
            <div className="container">
                <Link className="btn btn-light mb-2" to={`/topics/${level.topic}/`}>
                    <FontAwesomeIcon icon={faChevronLeft}/> Back to Topic Page
                </Link>
                <h1>{level.title} {level.playable ?  <span className="badge badge-success">Playable</span> : <span className="badge badge-secondary">Unplayable</span>}</h1>
                <p>{level.description}</p>
                <div className="mb-4">
                    {
                        level.playable
                            ? <Link className="btn btn-primary" to={`/topics/${level.topic}/levels/${level.id}/student-reports`}>
                                View Student Reports
                            </Link>
                            : <button className="btn btn-primary" onClick={this.onPublish}>
                                Publish Level
                            </button>
                    }
                </div>
                <h2>Learning Materials</h2>
                <LearningMaterialList levelId={level.id} editable={!level.playable} />
                <br />
                <h2>Questions</h2>
                <QuestionsList levelId={level.id} editable={!level.playable} />
            </div>
        );
    }
}

LevelPage.propTypes = {
    /** An object containing the topic ID and level ID based on which data is displayed */
    match: PropTypes.object.isRequired,

    /** A boolean to determine if the level is still being loaded by the `retrieveLevel` action creator (true: still loading, false: fully loaded) */
    levelLoading: PropTypes.bool.isRequired,
    /** A boolean to determine if the level failed to be loaded by the `retrieveLevel` action creator (true: still loading or failed to load, false: successful load) */
    levelFailed: PropTypes.bool,
    /** A level object loaded by the `retrieveLevel` action creator */
    level: PropTypes.object,

    /** An action creator for retrieving level name */
    retrieveLevel: PropTypes.func.isRequired,

};

const mapStateToProps = state => ({
    levelLoading: selectLevelLoading(state),
    levelFailed: selectLevelFailed(state),
    level: selectLevel(state),
});

const dispatchers = {
    retrieveLevel,
    publishLevel,
};


export default connect(mapStateToProps, dispatchers)(LevelPage);