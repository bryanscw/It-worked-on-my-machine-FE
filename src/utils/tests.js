import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from '../redux/rootReducer';

export const renderWithRedux = (component, initialState = {}) => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )
    return <Provider store={store}>{component}</Provider>;
}

export const renderWithRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return <Router history={history}>{component}</Router>
}

export const renderWithReduxRouter = (component, initialState = {}, route = '/') => {
    return renderWithRedux(renderWithRouter(component, route), initialState);
}