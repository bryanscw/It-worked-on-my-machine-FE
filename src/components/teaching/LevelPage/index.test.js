import React from 'react'
import { cleanup, waitForElement } from '@testing-library/react'
import axiosMock from 'axios'

import { renderWithReduxRouter } from '../../../utils/tests.js';
import LevelPage from './index.js';

jest.mock('axios');

beforeEach(() => {
    axiosMock.get.mockImplementation(url => {
        if (/topics/.test(url)) {
            return Promise.resolve({
                data: {
                    id: 1,
                    title: "Adding 1 to 100",
                    description: "Small numbers.",
                    playable: true,
                    topic: 1
                }
            });

        } else if (/learningMaterials/.test(url)) {
            return Promise.resolve({
                data: {
                    content: [
                        {
                            id: 1,
                            title: "Basic addition",
                            description: "Interesting video",
                            link: "https://www.youtube.com/embed/Fe8u2I3vmHU",
                        }
                    ]
                }
            });

        } else if (/questions/.test(url)) {
            return Promise.resolve({
                data: {
                    content: [
                        {
                            id: 1,
                            questionText: "1 + 1",
                            options: {
                                '0': 1,
                                '1': 2,
                            },
                            answer: 1,
                        }
                    ]
                }
            });
        }

        return Promise.reject(new Error('not found'));
    });
});

afterEach(() => {
    cleanup();
    jest.clearAllMocks();
});

const matchObject = {
    params: {
        topicId: 1,
        levelId: 1,
    }
}

it('should take a snapshot', async () => {
    const { asFragment, getByText } = renderWithReduxRouter(<LevelPage match={matchObject} />);
    await waitForElement(() => getByText('Learning Materials'));
    await waitForElement(() => getByText('Questions'));
    expect(asFragment()).toMatchSnapshot();
})

it('should load learning material and questions', () => {
    const { getByTestId } = renderWithReduxRouter(<LevelPage match={matchObject} />); 
    expect(getByTestId('loader')).toBeTruthy();
});

it('should load and display the learning material and questions', async () => {
    const { getByText } = renderWithReduxRouter(<LevelPage match={matchObject} />);
    const learningMaterial = await waitForElement(() => getByText('Learning Materials'));
    const questions = await waitForElement(() => getByText('Questions'));
    expect(axiosMock.get).toHaveBeenCalledTimes(3);
    expect(learningMaterial).toBeTruthy();
    expect(questions).toBeTruthy();
})

it('should not display edit or delete buttons and should display view student reports if playable', async () => {
    const { getByText, container } = renderWithReduxRouter(<LevelPage match={matchObject} />);
    
    await waitForElement(() => getByText('Learning Materials'));
    await waitForElement(() => getByText('Questions'));
    
    const editButton = container.querySelector('[data-icon="edit"]');
    expect(editButton).toBeFalsy();

    const deleteButton = container.querySelector('[data-icon="trash"]');
    expect(deleteButton).toBeFalsy();

    const viewStudentReports = await waitForElement(() => getByText('View Student Reports'));
    expect(viewStudentReports).toBeTruthy();
});

it('should display edit or delete buttons if unplayable', async () => {
    axiosMock.get.mockImplementation((url) => {
        if (/topics/.test(url)) {
            return Promise.resolve({
                data: {
                    id: 1,
                    title: "Adding 1 to 100",
                    description: "Small numbers.",
                    playable: false,
                    topic: 1
                }
            });

        } else if (/learningMaterials/.test(url)) {
            return Promise.resolve({
                data: {
                    content: [
                        {
                            id: 1,
                            title: "Basic addition",
                            description: "Interesting video",
                            link: "https://www.youtube.com/embed/Fe8u2I3vmHU",
                        }
                    ]
                }
            });

        } else if (/questions/.test(url)) {
            return Promise.resolve({
                data: {
                    content: [
                        {
                            id: 1,
                            questionText: "1 + 1",
                            options: {
                                '0': 1,
                                '1': 2,
                            },
                            answer: 1,
                        }
                    ]
                }
            });
        }

        return Promise.reject(new Error('not found'));
    });
    
    const { getByText, container, queryByText } = renderWithReduxRouter(<LevelPage match={matchObject} />);
    
    await waitForElement(() => getByText('Learning Materials'));
    await waitForElement(() => getByText('Questions'));
    
    const editButton = await waitForElement(() => container.querySelector('[data-icon="edit"]'));
    expect(editButton).toBeTruthy();

    const deleteButton = await waitForElement(() => container.querySelector('[data-icon="trash"]'));
    expect(deleteButton).toBeTruthy();

    const publishLevel = queryByText('Publish Level');
    expect(publishLevel).toBeTruthy();

    const viewStudentReports = queryByText('View Student Reports');
    expect(viewStudentReports).toBeFalsy();
});