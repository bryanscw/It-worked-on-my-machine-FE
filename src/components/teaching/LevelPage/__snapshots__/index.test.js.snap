// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <a
      class="btn btn-light mb-2"
      href="/topics/1/"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-left fa-w-10 "
        data-icon="chevron-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          fill="currentColor"
        />
      </svg>
       Back to Topic Page
    </a>
    <h1>
      Adding 1 to 100 
      <span
        class="badge badge-success"
      >
        Playable
      </span>
    </h1>
    <p>
      Smol numbers.
    </p>
    <div
      class="mb-4"
    >
      <a
        class="btn btn-primary"
        href="/topics/1/levels/1/student-reports"
      >
        View Student Reports
      </a>
    </div>
    <h2>
      Learning Materials
    </h2>
    <div
      class="row"
    >
      <div
        class="col-lg-6 mb-4"
      >
        <div
          class="card h-100"
        >
          <div
            class="video-box card-img-top"
          >
            <div>
              <iframe
                allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
                allowfullscreen=""
                frameborder="0"
                src="https://www.youtube.com/embed/Fe8u2I3vmHU"
                title="Learning Material Video"
              />
            </div>
          </div>
          <div
            class="card-body"
          >
            <h3
              class="card-title"
            >
              Basic addition
            </h3>
            <p
              class="card-text"
            >
              Interesting video
            </p>
          </div>
        </div>
      </div>
    </div>
    <br />
    <h2>
      Questions
    </h2>
    <div
      class="card-columns"
    >
      <div
        class="card"
      >
        <h3
          class="card-header"
        >
          1 + 1
        </h3>
        <ul
          class="list-group list-group-flush"
        >
          <li
            class="list-group-item"
          >
            1 
          </li>
          <li
            class="list-group-item"
          >
            2 
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-check fa-w-16 text-success"
              data-icon="check"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
                fill="currentColor"
              />
            </svg>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
